backendRoot/src/main/java/com/virtuslab/gitmachete/backend/root/GitMacheteRepositoryBuilder.java

```diff
@@ -114,55 +116,81 @@ private GitMacheteRootBranch createGitMacheteRootBranch(IGitCoreRepository gitCo
         .getOrElseThrow(e -> new GitMacheteException(e));
 
-    Optional<BaseGitCoreCommit> coreForkPoint = Try.of(() -> coreLocalBranch.deriveForkPoint())
-        .getOrElseThrow(e -> new GitMacheteException(e));
+    Optional<BaseGitCoreCommit> coreForkPoint = deduceForkPoint(ancestorityCache, coreLocalBranch,
+        parentEntryCoreLocalBranch);
 
     BaseGitCoreCommit corePointedCommit = Try.of(() -> coreLocalBranch.getPointedCommit())
         .getOrElseThrow(e -> new GitMacheteException(e));
 
-    if (!coreForkPoint.isPresent()) {
-      throw new GitMacheteException(String
-          .format("Cannot derive the fork point of non root branch \"%s\"", entry.getName()));
-    }
-
     // translate IGitCoreCommit list to IGitMacheteCommit list
-    List<IGitMacheteCommit> commits = Try.of(() -> coreLocalBranch.deriveCommitsUntil(coreForkPoint.get()))
-        .getOrElseThrow(e -> new GitMacheteException(e))
-        .map(GitMacheteCommit::new)
-        .collect(List.collector());
+    List<IGitMacheteCommit> commits = coreForkPoint.isPresent()
+        ? Try.of(() -> coreLocalBranch.deriveCommitsUntil(coreForkPoint.get()))
+            .getOrElseThrow(e -> new GitMacheteException(e))
+            .map(GitMacheteCommit::new)
+            .collect(List.collector())
+        : List.empty();
 
     var pointedCommit = new GitMacheteCommit(corePointedCommit);
-    var forkPoint = new GitMacheteCommit(coreForkPoint.get());
+    var forkPoint = coreForkPoint.isPresent() ? new GitMacheteCommit(coreForkPoint.get()) : null;
     var syncToOriginStatus = deriveSyncToOriginStatus(coreLocalBranch);
-    var syncToParentStatus = deriveSyncToParentStatus(gitCoreRepository, coreLocalBranch, parentEntryCoreLocalBranch,
-        coreForkPoint.get());
+    var syncToParentStatus = deriveSyncToParentStatus(ancestorityCache, coreLocalBranch, parentEntryCoreLocalBranch,
+        coreForkPoint.orElse(null));
     var customAnnotation = entry.getCustomAnnotation().orElse(null);
-    var subbranches = deriveDownstreamBranches(gitCoreRepository, coreLocalBranch, entry);
+    var subbranches = deriveDownstreamBranches(gitCoreRepository, ancestorityCache, coreLocalBranch, entry);
 
-    return new GitMacheteNonRootBranch(entry.getName(), subbranches, pointedCommit, forkPoint, commits,
-        syncToOriginStatus, syncToParentStatus, customAnnotation);
+    return new GitMacheteNonRootBranch(entry.getName(), subbranches, forkPoint, pointedCommit,
+        commits, syncToOriginStatus, syncToParentStatus, customAnnotation);
+  }
+
+  private Optional<BaseGitCoreCommit> deduceForkPoint(AncestorityCache ancestorityCache,
+      IGitCoreLocalBranch coreLocalBranch,
+      IGitCoreLocalBranch parentCoreLocalBranch) throws GitMacheteException {
+    var forkPointOptional = Try.of(() -> coreLocalBranch.deriveForkPoint())
+        .getOrElseThrow(e -> new GitMacheteException(e));
+    if (!forkPointOptional.isPresent()) {
```

Comment:
Shouldn't this be without `!` or I miss something? I think if fork point was found we should then return it.

Reply:
Hmmm yeah I also don't get this logic now that you pointed that out...

Reply:
Oh ok, if we found the FP(A) we still need to check if the FP(A) we found is descendant of parent(A)

Reply:
But what seems to be missing here is the fallback to parent(A) in case FP(A) is missing but A is descendant of parent(A) :thinking:

Reply:
done (test case added too)